More useful error messages if environment is not set up correctly or extra-files are missing

isofit 3.0 (current dev branch)

There are a couple of not very user-friendly exceptions that are raised in case the environment is not correctly configured or the ISOFIT extra-files are missing. I think these cases could easily be catched beforehand and a useful error message could be raised pointing to the solution.

<details> 
  <summary>FileNotFoundError if `isofit download data` was not executed -> data folder missing</summary>
    
```python
test_isofit_enmap.py::Test_ISOFIT_EnMAP::test_run_isofit 

======================== 1 failed, 3 warnings in 30.74s ========================
2024-03-14 20:17:49,255	INFO worker.py:1724 -- Started a local Ray instance.
Stopping ray.
FAILED
tests/test_isofit_enmap.py:63 (Test_ISOFIT_EnMAP.test_run_isofit)
self = <tests.test_isofit_enmap.Test_ISOFIT_EnMAP testMethod=test_run_isofit>

    def test_run_isofit(self):
        # from geoarray import GeoArray
        # gA = GeoArray('/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/data_in/emp20220712t184754_rdn_sub.bil')
        # gA.show()
    
        from time import time
        t0 = time()
    
>       IsofitEnMAP()._apply_oe(
            # input_radiance='/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/data_in/emp20220712t184754_rdn_sub.bil',
            input_radiance='/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/data_in/ENMAP01-____L1X-DT000000XXXX_20220712T000000Z_00x_VXXXXXX_XXXXXXTXXXXXXZ.bil',
            input_loc='/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/data_in/emp20220712t184754_loc_sub.bil',
            input_obs='/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/data_in/emp20220712t184754_obs_sub.bil',
            working_directory='/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/data_out/',
            surface_path='/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/surface/surface_20221020_EnMAP.mat',
            wavelength_path='/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/sensor_new/enmap_wavelengths.txt',
            log_file='/home/gfz-fe/scheffler/temp/EnPT/isofit_implementation/data_out/isofit.log',
            presolve=True,
            emulator_base='/home/gfz-fe/scheffler/sRTMnet_v100/sRTMnet_v100',  # FIXME why not /home/gfz-fe/scheffler/sRTMnet_v100/
            n_cores=30
        )

test_isofit_enmap.py:72: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../enpt/processors/atmospheric_correction/_isofit_enmap.py:108: in _apply_oe
    apply_oe(SimpleNamespace(**params))
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/utils/apply_oe.py:590: in apply_oe
    retrieval_full.run()
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/core/isofit.py:182: in run
    res = list(
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/ray/util/actor_pool.py:170: in get_generator
    yield self.get_next_unordered()
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/ray/util/actor_pool.py:370: in get_next_unordered
    return ray.get(future)
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/ray/_private/auto_init_hook.py:22: in auto_init_wrapper
    return fn(*args, **kwargs)
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/ray/_private/client_mode_hook.py:103: in wrapper
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

object_refs = [ObjectRef(8c4854248414f6339358be101ae77cb2af9663a30100000001000000)]

    @PublicAPI
    @client_mode_hook
    def get(
        object_refs: Union["ObjectRef[Any]", Sequence["ObjectRef[Any]"]],
        *,
        timeout: Optional[float] = None,
    ) -> Union[Any, List[Any]]:
        """Get a remote object or a list of remote objects from the object store.
    
        This method blocks until the object corresponding to the object ref is
        available in the local object store. If this object is not in the local
        object store, it will be shipped from an object store that has it (once the
        object has been created). If object_refs is a list, then the objects
        corresponding to each object in the list will be returned.
    
        Ordering for an input list of object refs is preserved for each object
        returned. That is, if an object ref to A precedes an object ref to B in the
        input list, then A will precede B in the returned list.
    
        This method will issue a warning if it's running inside async context,
        you can use ``await object_ref`` instead of ``ray.get(object_ref)``. For
        a list of object refs, you can use ``await asyncio.gather(*object_refs)``.
    
        Passing :class:`~ObjectRefGenerator` is not allowed.
    
        Related patterns and anti-patterns:
    
        - :doc:`/ray-core/patterns/ray-get-loop`
        - :doc:`/ray-core/patterns/unnecessary-ray-get`
        - :doc:`/ray-core/patterns/ray-get-submission-order`
        - :doc:`/ray-core/patterns/ray-get-too-many-objects`
    
    
        Args:
            object_refs: Object ref of the object to get or a list of object refs
                to get.
            timeout (Optional[float]): The maximum amount of time in seconds to
                wait before returning. Set this to None will block until the
                corresponding object becomes available. Setting ``timeout=0`` will
                return the object immediately if it's available, else raise
                GetTimeoutError in accordance with the above docstring.
    
        Returns:
            A Python object or a list of Python objects.
    
        Raises:
            GetTimeoutError: A GetTimeoutError is raised if a timeout is set and
                the get takes longer than timeout to return.
            Exception: An exception is raised if the task that created the object
                or that created one of the objects raised an exception.
        """
        worker = global_worker
        worker.check_connected()
    
        if hasattr(worker, "core_worker") and worker.core_worker.current_actor_is_asyncio():
            global blocking_get_inside_async_warned
            if not blocking_get_inside_async_warned:
                logger.warning(
                    "Using blocking ray.get inside async actor. "
                    "This blocks the event loop. Please use `await` "
                    "on object ref with asyncio.gather if you want to "
                    "yield execution to the event loop instead."
                )
                blocking_get_inside_async_warned = True
    
        with profiling.profile("ray.get"):
            # TODO(sang): Should make ObjectRefGenerator
            # compatible to ray.get for dataset.
            if isinstance(object_refs, ObjectRefGenerator):
                return object_refs
    
            is_individual_id = isinstance(object_refs, ray.ObjectRef)
            if is_individual_id:
                object_refs = [object_refs]
    
            if not isinstance(object_refs, list):
                raise ValueError(
                    f"Invalid type of object refs, {type(object_refs)}, is given. "
                    "'object_refs' must either be an ObjectRef or a list of ObjectRefs. "
                )
    
            # TODO(ujvl): Consider how to allow user to retrieve the ready objects.
            values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
            for i, value in enumerate(values):
                if isinstance(value, RayError):
                    if isinstance(value, ray.exceptions.ObjectLostError):
                        worker.core_worker.dump_object_store_memory_usage()
                    if isinstance(value, RayTaskError):
>                       raise value.as_instanceof_cause()
E                       ray.exceptions.RayTaskError(FileNotFoundError): ray::Worker.run_set_of_spectra() (pid=1671034, ip=139.17.56.61, actor_id=9358be101ae77cb2af9663a301000000, repr=<isofit.core.isofit.Worker object at 0x7effdcfd8350>)
E                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E                         File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/core/isofit.py", line 250, in run_set_of_spectra
E                           input_data = self.io.get_components_at_index(row, col)
E                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E                         File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/core/fileio.py", line 495, in get_components_at_index
E                           geom = Geometry(
E                                  ^^^^^^^^^
E                         File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/core/geometry.py", line 61, in __init__
E                           self.earth_sun_distance_reference = np.loadtxt(self.earth_sun_distance_path)
E                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E                         File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/numpy/lib/npyio.py", line 1373, in loadtxt
E                           arr = _read(fname, dtype=dtype, comment=comment, delimiter=delimiter,
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E                         File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/numpy/lib/npyio.py", line 992, in _read
E                           fh = np.lib._datasource.open(fname, 'rt', encoding=encoding)
E                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E                         File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/numpy/lib/_datasource.py", line 193, in open
E                           return ds.open(path, mode, encoding=encoding, newline=newline)
E                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E                         File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/numpy/lib/_datasource.py", line 533, in open
E                           raise FileNotFoundError(f"{path} not found.")
E                       FileNotFoundError: /home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/data/earth_sun_distance.txt not found.
```
</details>


<details> 
  <summary>FileNotFoundError if `isofit download examples` was not executed (examples folder missing) </summary>

```python
../enpt/processors/atmospheric_correction/_isofit_enmap.py:100: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/utils/apply_oe.py:490: in apply_oe
    retrieval_h2o.run()
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/core/isofit.py:119: in run
    self.fm = fm = ForwardModel(self.config)
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/core/forward.py:80: in __init__
    self.RT = RadiativeTransfer(self.full_config)
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/radiative_transfer.py:106: in __init__
    rte = RTE[confRT.engine_name](**params)
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/radiative_transfer_engine.py:218: in __init__
    self.runSimulations()
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/radiative_transfer_engine.py:416: in runSimulations
    pre = self.preSim()
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/sRTMnet.py:100: in preSim
    sim = SixSRT(
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/six_s.py:73: in __init__
    super().__init__(engine_config, **kwargs)
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/radiative_transfer_engine.py:218: in __init__
    self.runSimulations()
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/radiative_transfer_engine.py:442: in runSimulations
    post = self.postSim()
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/six_s.py:159: in postSim
    irr = np.loadtxt(self.engine_config.irradiance_file, comments="#")
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/numpy/lib/npyio.py:1373: in loadtxt
    arr = _read(fname, dtype=dtype, comment=comment, delimiter=delimiter,
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/numpy/lib/npyio.py:992: in _read
    fh = np.lib._datasource.open(fname, 'rt', encoding=encoding)
../../../../mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/numpy/lib/_datasource.py:193: in open
    return ds.open(path, mode, encoding=encoding, newline=newline)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <numpy.DataSource object at 0x7f7d31e4d210>
path = '/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/examples/20151026_SantaMonica/data/prism_optimized_irr.dat'
mode = 'rt', encoding = None, newline = None

    def open(self, path, mode='r', encoding=None, newline=None):
        """
        Open and return file-like object.
    
        If `path` is an URL, it will be downloaded, stored in the
        `DataSource` directory and opened from there.
    
        Parameters
        ----------
        path : str
            Local file path or URL to open.
        mode : {'r', 'w', 'a'}, optional
            Mode to open `path`.  Mode 'r' for reading, 'w' for writing,
            'a' to append. Available modes depend on the type of object
            specified by `path`. Default is 'r'.
        encoding : {None, str}, optional
            Open text file with given encoding. The default encoding will be
            what `io.open` uses.
        newline : {None, str}, optional
            Newline to use when reading text file.
    
        Returns
        -------
        out : file object
            File object.
    
        """
    
        # TODO: There is no support for opening a file for writing which
        #       doesn't exist yet (creating a file).  Should there be?
    
        # TODO: Add a ``subdir`` parameter for specifying the subdirectory
        #       used to store URLs in self._destpath.
    
        if self._isurl(path) and self._iswritemode(mode):
            raise ValueError("URLs are not writeable")
    
        # NOTE: _findfile will fail on a new file opened for writing.
        found = self._findfile(path)
        if found:
            _fname, ext = self._splitzipext(found)
            if ext == 'bz2':
                mode.replace("+", "")
            return _file_openers[ext](found, mode=mode,
                                      encoding=encoding, newline=newline)
        else:
>           raise FileNotFoundError(f"{path} not found.")
E           FileNotFoundError: /home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/examples/20151026_SantaMonica/data/prism_optimized_irr.dat not found.
```
</details>

<details> 
  <summary>NameError if SIXS_DIR not set to /path/to/6sV2.1 </summary>

```python
[...]
:task_name:streamSimulation
[2024-03-14 17:31:55,918 E 1580751 1580751] core_worker.cc:1749: Pushed Error with JobID: 01000000 of type: task with message: ray::streamSimulation() (pid=1580751, ip=139.17.56.61)
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/radiative_transfer_engine.py", line 624, in streamSimulation
    data = reader(point)
           ^^^^^^^^^^^^^
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/six_s.py", line 151, in readSim
    return self.parse_file(file=file, wl=self.wl, wl_size=self.wl.size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/six_s.py", line 445, in parse_file
    Logger.error(f"Failed to parse any data for point: {point}")
                                                        ^^^^^
NameError: name 'point' is not defined at time: 1.71043e+09
```
</details>

<details> 
  <summary>FileNotFoundError if EMULATOR_PATH not set or apply_oe(emulator_base=None) </summary>

```python
:task_name:streamSimulation
[2024-03-14 13:50:38,732 E 1079104 1079104] core_worker.cc:1749: Pushed Error with JobID: 01000000 of type: task with message: ray::streamSimulation() (pid=1079104, ip=139.17.56.61)
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/radiative_transfer_engine.py", line 624, in streamSimulation
    data = reader(point)
           ^^^^^^^^^^^^^
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/modtran.py", line 270, in readSim
    solzen = self.load_tp6(f"{file}.tp6")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/modtran.py", line 200, in load_tp6
    with open(file, "r") as tp6:
         ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpfm0mwnjo/lut_full/AERFRAC_2-0.6004_H2OSTR-1.7825_surface_elevation_km-1.0528.tp6' at time: 1.71042e+09
:task_name:streamSimulation
:task_name:streamSimulation
[2024-03-14 13:50:38,885 E 1079104 1079104] core_worker.cc:1749: Pushed Error with JobID: 01000000 of type: task with message: ray::streamSimulation() (pid=1079104, ip=139.17.56.61)
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/radiative_transfer_engine.py", line 624, in streamSimulation
    data = reader(point)
           ^^^^^^^^^^^^^
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/modtran.py", line 270, in readSim
    solzen = self.load_tp6(f"{file}.tp6")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gfz-fe/mambaforge/envs/enpt_full_dev/lib/python3.11/site-packages/isofit/radiative_transfer/modtran.py", line 200, in load_tp6
    with open(file, "r") as tp6:
         ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpfm0mwnjo/lut_full/AERFRAC_2-0.6004_H2OSTR-1.7825_surface_elevation_km-1.2602.tp6' at time: 1.71042e+09
```
</details>

--- a/isofit/core/fileio.py
+++ b/isofit/core/fileio.py
@@ -762,7 +762,7 @@ class IO:
             Loaded ESD. If the file fails to load, creates a default
         """
         if file is None:
-            file = os.path.join(env.data, "earth_sun_distance.txt")
+            file = env.path("data", "earth_sun_distance.txt")
 
         try:
             esd = np.loadtxt(file)

--- a/isofit/data/cli/__init__.py
+++ b/isofit/data/cli/__init__.py
@@ -43,12 +43,19 @@ def validate_all():
 def env_validate(keys, **kwargs):
     """
     Utility function for the `env` object to quickly validate specific dependencies
+
+    Parameters
+    ----------
+    keys : list
+        List of validator functions to call
     """
+    error = kwargs.get("error", print)
+
     all_valid = True
     for key in keys:
         module = Modules.get(key)
         if module is None:
-            print(f"Product not found: {key}")
+            error(f"Product not found: {key}")
             all_valid = False
         else:
             all_valid &= module.validate(**kwargs)

--- a/isofit/data/cli/data.py
+++ b/isofit/data/cli/data.py
@@ -87,7 +87,7 @@ def validate(path=None, debug=print, error=print, **_):
 
     if not (path := Path(path)).exists():
         error(
-            "Error: Path does not exist, please download it via `isofit download data`"
+            "Error: Data path does not exist, please download it via `isofit download data`"
         )
         return False
 

--- a/isofit/data/cli/data.py
+++ b/isofit/data/cli/data.py
@@ -59,7 +59,7 @@ def download_cli(**kwargs):
     download(**kwargs)
 
 
-def validate(path=None, **_):
+def validate(path=None, debug=print, error=print, **_):
     """
     Validates an ISOFIT data installation
 
@@ -67,6 +67,10 @@ def validate(path=None, **_):
     ----------
     path : str, default=None
         Path to verify. If None, defaults to the ini path
+    debug : function, default=print
+        Print function to use for debug messages, eg. logging.debug
+    error : function, default=print
+        Print function to use for error messages, eg. logging.error
     **_ : dict
         Ignores unused params that may be used by other validate functions. This is to
         maintain compatibility with env.validate
@@ -79,10 +83,10 @@ def validate(path=None, **_):
     if path is None:
         path = env.data
 
-    print(f"Verifying path for ISOFIT data: {path}")
+    debug(f"Verifying path for ISOFIT data: {path}")
 
     if not (path := Path(path)).exists():
-        print(
+        error(
             "Error: Path does not exist, please download it via `isofit download data`"
         )
         return False
@@ -95,12 +99,12 @@ def validate(path=None, **_):
     ]
     files = list(path.glob("*"))
     if not all([path / file in files for file in check]):
-        print(
+        error(
             "Error: ISOFIT data do not appear to be installed correctly, please ensure it is"
         )
         return False
 
-    print("Path is valid")
+    debug("Path is valid")
     return True
 
 

--- a/isofit/data/cli/examples.py
+++ b/isofit/data/cli/examples.py
@@ -61,7 +61,7 @@ def download_cli(**kwargs):
     download(**kwargs)
 
 
-def validate(path=None, **_):
+def validate(path=None, debug=print, error=print, **_):
     """
     Validates an ISOFIT examples installation
 
@@ -69,6 +69,10 @@ def validate(path=None, **_):
     ----------
     path : str, default=None
         Path to verify. If None, defaults to the ini path
+    debug : function, default=print
+        Print function to use for debug messages, eg. logging.debug
+    error : function, default=print
+        Print function to use for error messages, eg. logging.error
     **_ : dict
         Ignores unused params that may be used by other validate functions. This is to
         maintain compatibility with env.validate
@@ -81,10 +85,10 @@ def validate(path=None, **_):
     if path is None:
         path = env.examples
 
-    print(f"Verifying path for ISOFIT examples: {path}")
+    debug(f"Verifying path for ISOFIT examples: {path}")
 
     if not (path := Path(path)).exists():
-        print(
+        error(
             "Error: Path does not exist, please download it via `isofit download examples`"
         )
         return False
@@ -100,12 +104,12 @@ def validate(path=None, **_):
         "py-hypertrace",
     ]
     if not list(path.glob("*")) != expected:
-        print(
+        error(
             "Error: ISOFIT examples do not appear to be installed correctly, please ensure it is"
         )
         return False
 
-    print("Path is valid")
+    debug("Path is valid")
     return True
 
 

--- a/isofit/data/cli/examples.py
+++ b/isofit/data/cli/examples.py
@@ -89,7 +89,7 @@ def validate(path=None, debug=print, error=print, **_):
 
     if not (path := Path(path)).exists():
         error(
-            "Error: Path does not exist, please download it via `isofit download examples`"
+            "Error: Examples path does not exist, please download it via `isofit download examples`"
         )
         return False
 

--- a/isofit/data/cli/image_cube.py
+++ b/isofit/data/cli/image_cube.py
@@ -77,7 +77,7 @@ def download_cli(**kwargs):
     download(**kwargs)
 
 
-def validate(path=None, size="both", **_):
+def validate(path=None, size="both", debug=print, error=print, **_):
     """
     Validates an ISOFIT image cube data installation
 
@@ -87,6 +87,10 @@ def validate(path=None, size="both", **_):
         Path to verify. If None, defaults to the ini path
     size : "both" | "small" | "medium"
         Which chunk size to validate
+    debug : function, default=print
+        Print function to use for debug messages, eg. logging.debug
+    error : function, default=print
+        Print function to use for error messages, eg. logging.error
     **_ : dict
         Ignores unused params that may be used by other validate functions. This is to
         maintain compatibility with env.validate
@@ -102,19 +106,19 @@ def validate(path=None, size="both", **_):
     if path is None:
         path = Path(env.imagecube)
 
-    print(f"Verifying path for ISOFIT {size} image cube: {path}")
+    debug(f"Verifying path for ISOFIT {size} image cube: {path}")
 
     sizes = {"small": "7000-7010", "medium": "7k-8k"}
     for kind in ("loc", "obs", "rdn"):
         file = path / size / f"ang20170323t202244_{kind}_{sizes[size]}"
         if not file.exists():
-            print(
+            error(
                 f"Error: ISOFIT {size} image cube data do not appear to be installed correctly, please ensure it is"
             )
-            print(f"Missing file: {file}")
+            error(f"Missing file: {file}")
             return False
 
-    print("Path is valid")
+    debug("Path is valid")
     return True
 
 

--- a/isofit/data/cli/sixs.py
+++ b/isofit/data/cli/sixs.py
@@ -107,7 +107,9 @@ def validate(path=None, debug=print, error=print, **_):
     debug(f"Verifying path for 6S: {path}")
 
     if not (path := Path(path)).exists():
-        error("Error: Path does not exist, please download it via `isofit download 6S`")
+        error(
+            "Error: 6S path does not exist, please download it via `isofit download 6S`"
+        )
         return False
 
     if not (path / f"sixsV2.1").exists():

--- a/isofit/data/cli/sixs.py
+++ b/isofit/data/cli/sixs.py
@@ -80,7 +80,7 @@ def download_cli(**kwargs):
     download(**kwargs)
 
 
-def validate(path=None, **_):
+def validate(path=None, debug=print, error=print, **_):
     """
     Validates a 6S installation
 
@@ -88,6 +88,10 @@ def validate(path=None, **_):
     ----------
     path : str, default=None
         Path to verify. If None, defaults to the ini path
+    debug : function, default=print
+        Print function to use for debug messages, eg. logging.debug
+    error : function, default=print
+        Print function to use for error messages, eg. logging.error
     **_ : dict
         Ignores unused params that may be used by other validate functions. This is to
         maintain compatibility with env.validate
@@ -100,19 +104,19 @@ def validate(path=None, **_):
     if path is None:
         path = env.sixs
 
-    print(f"Verifying path for 6S: {path}")
+    debug(f"Verifying path for 6S: {path}")
 
     if not (path := Path(path)).exists():
-        print("Error: Path does not exist, please download it via `isofit download 6S`")
+        error("Error: Path does not exist, please download it via `isofit download 6S`")
         return False
 
     if not (path / f"sixsV2.1").exists():
-        print(
+        error(
             "Error: 6S does not appear to be installed correctly, please ensure it is"
         )
         return False
 
-    print("Path is valid")
+    debug("Path is valid")
     return True
 
 

--- a/isofit/data/cli/srtmnet.py
+++ b/isofit/data/cli/srtmnet.py
@@ -88,7 +88,7 @@ def download_cli(**kwargs):
     download(**kwargs)
 
 
-def validate(path=None, **_):
+def validate(path=None, debug=print, error=print, **_):
     """
     Validates an sRTMnet installation
 
@@ -96,6 +96,10 @@ def validate(path=None, **_):
     ----------
     path : str, default=None
         Path to verify. If None, defaults to the ini path
+    debug : function, default=print
+        Print function to use for debug messages, eg. logging.debug
+    error : function, default=print
+        Print function to use for error messages, eg. logging.error
     **_ : dict
         Ignores unused params that may be used by other validate functions. This is to
         maintain compatibility with env.validate
@@ -108,23 +112,23 @@ def validate(path=None, **_):
     if path is None:
         path = env.srtmnet
 
-    print(f"Verifying path for sRTMnet: {path}")
+    debug(f"Verifying path for sRTMnet: {path}")
 
     if not (path := Path(path)).exists():
-        print(
+        error(
             "Error: sRTMnet path does not exist, please download it via `isofit download sRTMnet`"
         )
         return False
 
     if not list(path.glob("*.h5")):
-        print("Error: sRTMnet model not found, please download it")
+        error("Error: sRTMnet model not found, please download it")
         return False
 
     if not list(path.glob("*_aux.npz")):
-        print("Error: sRTMnet aux file not found, please download it")
+        error("Error: sRTMnet aux file not found, please download it")
         return False
 
-    print("Path is valid")
+    debug("Path is valid")
     return True
 
 

--- a/isofit/data/ini.py
+++ b/isofit/data/ini.py
@@ -195,7 +195,7 @@ class Ini:
         """
         self.validate([dir], debug=Logger.debug, error=Logger.error)
 
-        path = Path([self[dir], *path]).resolve()
+        path = Path(*[self[dir], *path]).resolve()
 
         if not path.exists():
             Logger.error(

--- a/isofit/data/ini.py
+++ b/isofit/data/ini.py
@@ -177,11 +177,45 @@ class Ini:
             except:
                 Logger.exception(f"Failed to dump ini to file: {self.ini}")
 
+    def path(self, dir: str, path: str) -> Path:
+        """
+        Retrieves a path under one of the env directories and validates the path exists.
+
+        Parameters
+        ----------
+        dir : str
+            One of the env directories, eg. "data", "examples"
+        path : str
+            Path to a file under the `dir`
+
+        Returns
+        -------
+        pathlib.Path
+            Validated full path
+        """
+        self.validate([dir], debug=Logger.debug, error=Logger.error)
+
+        path = (Path(self[dir]) / path).resolve()
+
+        if not path.exists():
+            Logger.error(
+                f"The following path does not exist, please verify your installation environment: {path}"
+            )
+
+        return path
+
     @staticmethod
     def validate(keys: List) -> bool:
         """
-        Validates known products. This function is defined by isofit.data.cli.__init__.py
+        Validates known products.
+
+        Parameters
+        ----------
+        keys : list
+            List of products to validate
         """
-        raise NotImplemented(
+        # Should never be raised as this function is defined and set in isofit.data.cli.__init__
+        # If this is hit, there's a critical environment issue
+        raise NotImplementedError(
             "ISOFIT failed to attach the validation function to this object"
         )

--- a/isofit/data/ini.py
+++ b/isofit/data/ini.py
@@ -177,7 +177,7 @@ class Ini:
             except:
                 Logger.exception(f"Failed to dump ini to file: {self.ini}")
 
-    def path(self, dir: str, path: str) -> Path:
+    def path(self, dir: str, *path: list) -> Path:
         """
         Retrieves a path under one of the env directories and validates the path exists.
 
@@ -185,7 +185,7 @@ class Ini:
         ----------
         dir : str
             One of the env directories, eg. "data", "examples"
-        path : str
+        *path : List[str]
             Path to a file under the `dir`
 
         Returns
@@ -195,7 +195,7 @@ class Ini:
         """
         self.validate([dir], debug=Logger.debug, error=Logger.error)
 
-        path = (Path(self[dir]) / path).resolve()
+        path = Path([self[dir], *path]).resolve()
 
         if not path.exists():
             Logger.error(

--- a/isofit/test/test_cli.py
+++ b/isofit/test/test_cli.py
@@ -40,12 +40,13 @@ def surface(cwd):
     # Generate the surface.mat using the image_cube example config
     # fmt: off
     surface_model(
-        config_path=f"{env.examples}/20171108_Pasadena/configs/ang20171108t184227_surface.json",
-        wavelength_path=f"{env.examples}/20171108_Pasadena/remote/20170320_ang20170228_wavelength_fit.txt",
+        config_path=env.path("examples", "20171108_Pasadena", "remote", "ang20171108t184227_surface.json"),
+        wavelength_path=env.path("examples", "20171108_Pasadena", "remote", "20170320_ang20170228_wavelength_fit.txt"),
         output_path=outp
     )
     # fmt: on
     # Return the path to the mat file
+
     return outp
 
 
@@ -89,7 +90,7 @@ def test_apply_oe(files, args, surface):
     ray.shutdown()
     sleep(120)
 
-    args[3] = os.path.join(env.srtmnet, "sRTMnet_v120.h5")
+    args[3] = env.path("srtmnet", "sRTMnet_v120.h5")
     arguments = ["apply_oe", *files, *args, "--surface_path", surface]
 
     # Passing non-string arguments to click is not allowed.

--- a/isofit/test/test_cli.py
+++ b/isofit/test/test_cli.py
@@ -40,7 +40,7 @@ def surface(cwd):
     # Generate the surface.mat using the image_cube example config
     # fmt: off
     surface_model(
-        config_path=env.path("examples", "20171108_Pasadena", "remote", "ang20171108t184227_surface.json"),
+        config_path=env.path("examples", "20171108_Pasadena", "configs", "ang20171108t184227_surface.json"),
         wavelength_path=env.path("examples", "20171108_Pasadena", "remote", "20170320_ang20170228_wavelength_fit.txt"),
         output_path=outp
     )

--- a/isofit/test/test_examples.py
+++ b/isofit/test/test_examples.py
@@ -23,7 +23,7 @@ from isofit.utils import surface_model
 def test_santa_monica(args, monkeypatch):
     """Run the Santa Monica test dataset."""
 
-    monkeypatch.chdir(f"{env.examples}/20151026_SantaMonica/")
+    monkeypatch.chdir(env.path("examples", "20151026_SantaMonica/"))
     surface_model("configs/prm20151026t173213_surface_coastal.json")
 
     runner = CliRunner()
@@ -47,7 +47,7 @@ def test_santa_monica(args, monkeypatch):
 def test_pasadena_modtran(args, monkeypatch):
     """Run Pasadena example dataset."""
 
-    monkeypatch.chdir(f"{env.examples}/20171108_Pasadena/")
+    monkeypatch.chdir(env.path("examples", "20171108_Pasadena/"))
     surface_model("configs/ang20171108t184227_surface.json")
 
     runner = CliRunner()
@@ -63,7 +63,7 @@ def test_pasadena_modtran(args, monkeypatch):
 def test_pasadena_topoflux(monkeypatch):
     """Run Pasadena topoflux example dataset."""
 
-    monkeypatch.chdir(f"{env.examples}/20171108_Pasadena/")
+    monkeypatch.chdir(env.path("examples", "20171108_Pasadena/"))
     surface_model("configs/ang20171108t184227_surface.json")
 
     model = Isofit(
@@ -77,7 +77,7 @@ def test_pasadena_topoflux(monkeypatch):
 def test_modtran_one(monkeypatch):
     """Run MODTRAN example dataset."""
 
-    monkeypatch.chdir(f"{env.examples}/20190806_ThermalIR/")
+    monkeypatch.chdir(env.path("examples", "20190806_ThermalIR/"))
     surface_model("configs/surface.json")
 
     model = Isofit(
@@ -91,7 +91,7 @@ def test_modtran_one(monkeypatch):
 # def test_profiling_cube_small(monkeypatch):
 #     """Run profiling datasets."""
 #
-#     monkeypatch.chdir(f"{env.examples}/profiling_cube/")
+#     monkeypatch.chdir(env.path("examples", "profiling_cube/"))
 #
 #     environ = os.environ.copy()
 #     environ["ISOFIT_DEBUG"] = "1"

--- a/isofit/utils/template_construction.py
+++ b/isofit/utils/template_construction.py
@@ -182,33 +182,29 @@ class Pathnames:
         self.sixs_path = os.getenv("SIXS_DIR", env.sixs)
 
         if args.sensor == "avcl":
-            self.noise_path = join(env.data, "avirisc_noise.txt")
+            self.noise_path = env.path("data", "avirisc_noise.txt")
         elif args.sensor == "emit":
-            self.noise_path = join(env.data, "emit_noise.txt")
+            self.noise_path = env.path("data", "emit_noise.txt")
             if self.input_channelized_uncertainty_path is None:
-                self.input_channelized_uncertainty_path = join(
-                    env.data, "emit_osf_uncertainty.txt"
+                self.input_channelized_uncertainty_path = env.path(
+                    "data", "emit_osf_uncertainty.txt"
                 )
             if self.input_model_discrepancy_path is None:
-                self.input_model_discrepancy_path = join(
-                    env.data, "emit_model_discrepancy.mat"
+                self.input_model_discrepancy_path = env.path(
+                    "data", "emit_model_discrepancy.mat"
                 )
+
         else:
             self.noise_path = None
             logging.info("no noise path found, proceeding without")
             # quit()
 
-        self.earth_sun_distance_path = abspath(join(env.data, "earth_sun_distance.txt"))
-        self.irradiance_file = abspath(
-            join(
-                env.examples,
-                "20151026_SantaMonica",
-                "data",
-                "prism_optimized_irr.dat",
-            )
+        self.earth_sun_distance_path = env.path("data", "earth_sun_distance.txt")
+        self.irradiance_file = env.path(
+            "examples", "20151026_SantaMonica", "data", "prism_optimized_irr.dat"
         )
 
-        self.aerosol_tpl_path = join(env.data, "aerosol_template.json")
+        self.aerosol_tpl_path = env.path("data", "aerosol_template.json")
         self.rdn_factors_path = None
         if args.rdn_factors_path is not None:
             self.rdn_factors_path = abspath(args.rdn_factors_path)
@@ -1143,7 +1139,7 @@ def load_climatology(
 
     """
 
-    aerosol_model_path = os.path.join(env.data, "aerosol_model.txt")
+    aerosol_model_path = env.path("data", "aerosol_model.txt")
     aerosol_state_vector = {}
     aerosol_lut_grid = {}
     aerosol_lut_ranges = [

--- a/isofit/utils/template_construction.py
+++ b/isofit/utils/template_construction.py
@@ -182,16 +182,16 @@ class Pathnames:
         self.sixs_path = os.getenv("SIXS_DIR", env.sixs)
 
         if args.sensor == "avcl":
-            self.noise_path = env.path("data", "avirisc_noise.txt")
+            self.noise_path = str(env.path("data", "avirisc_noise.txt"))
         elif args.sensor == "emit":
-            self.noise_path = env.path("data", "emit_noise.txt")
+            self.noise_path = str(env.path("data", "emit_noise.txt"))
             if self.input_channelized_uncertainty_path is None:
-                self.input_channelized_uncertainty_path = env.path(
-                    "data", "emit_osf_uncertainty.txt"
+                self.input_channelized_uncertainty_path = str(
+                    env.path("data", "emit_osf_uncertainty.txt")
                 )
             if self.input_model_discrepancy_path is None:
-                self.input_model_discrepancy_path = env.path(
-                    "data", "emit_model_discrepancy.mat"
+                self.input_model_discrepancy_path = str(
+                    env.path("data", "emit_model_discrepancy.mat")
                 )
 
         else:
@@ -199,12 +199,14 @@ class Pathnames:
             logging.info("no noise path found, proceeding without")
             # quit()
 
-        self.earth_sun_distance_path = env.path("data", "earth_sun_distance.txt")
-        self.irradiance_file = env.path(
-            "examples", "20151026_SantaMonica", "data", "prism_optimized_irr.dat"
+        self.earth_sun_distance_path = str(env.path("data", "earth_sun_distance.txt"))
+        self.irradiance_file = str(
+            env.path(
+                "examples", "20151026_SantaMonica", "data", "prism_optimized_irr.dat"
+            )
         )
 
-        self.aerosol_tpl_path = env.path("data", "aerosol_template.json")
+        self.aerosol_tpl_path = str(env.path("data", "aerosol_template.json"))
         self.rdn_factors_path = None
         if args.rdn_factors_path is not None:
             self.rdn_factors_path = abspath(args.rdn_factors_path)
@@ -1139,7 +1141,7 @@ def load_climatology(
 
     """
 
-    aerosol_model_path = env.path("data", "aerosol_model.txt")
+    aerosol_model_path = str(env.path("data", "aerosol_model.txt"))
     aerosol_state_vector = {}
     aerosol_lut_grid = {}
     aerosol_lut_ranges = [

